posthog/queries/trends/sql.py

```diff
@@ -389,62 +389,49 @@
 ON e.distinct_id = ep.distinct_id where team_id = %(team_id)s {event_filter} {filters} {parsed_date_from} {parsed_date_to} {actions_query} {null_person_filter}
 """
 
-_LIFECYCLE_EVENTS_QUERY = """
+LIFECYCLE_EVENTS_QUERY = """
 SELECT
-    person_id,
-
-    /*
-        We want to put the status of each period onto it's own line, so we
-        can easily aggregate over them. With the inner query we end up with a structure like:
+    {person_column} as person_id,
+    arraySort(groupUniqArray(dateTrunc(%(interval)s, toTimeZone(toDateTime(events.timestamp, %(timezone)s), %(timezone)s)))) AS all_activity,
+    arrayPopBack(arrayPushFront(all_activity, dateTrunc(%(interval)s, toTimeZone(toDateTime(created_at, %(timezone)s), %(timezone)s)))) as previous_activity,
+    arrayPopFront(arrayPushBack(all_activity, dateTrunc(%(interval)s, toDateTime('1970-01-01')))) as following_activity,
+    arrayMap((previous,current) -> if(
```

Comment:
For a potential future optimisation, we can reduce the number of intermediary arrays we generate here, which should reduce the memory consumption a bit

<img width="605" alt="image" src="https://user-images.githubusercontent.com/7115141/212476135-3f0690b1-c04b-497d-b89f-649734d53a90.png">

Reply:
My initial comment here RE memory is: have we tested this query using our biggest customer's data? And as a follow-up, using their most frequent event?

My recent experience with arrays tells me memory usage grows real fast so would be good to know if we blow up in those scenarios. (note that that might be ok if the current query for example times out in those scenarios)

(side note: could I get a link to the Metabase query?)

Reply:
I'm pretty sure the existing query doesn't work in that case, so it's not / shouldn't be objectively worse for them.

Reply:
good shout to test it for them though, please do it! 😋

Reply:
Did some quick testing and seems all is ok on that front, nice work!

Reply:
The size of the array for each person is bounded above by number of periods, so I expect the only case where it's really unfeasible is a year long timerange, and hourly intervals.

Reply:
Thanks for testing!